File: sdks/local-rp/dart/lib/src/crypto/crypto.dart
// package:cryptography_plus-backed crypto primitives this SDK needs
// (design doc, Language Crypto Matrix, "Dart" row): Ed25519 sign/verify,
// X25519 ECDH, AES-256-GCM / ChaCha20-Poly1305 AEAD, SHA-256 fingerprinting,
// HKDF-SHA256, and CSPRNG bytes.
//
// See the package README for the health-check evidence behind choosing
// `cryptography_plus` over the original `cryptography` (dint-dev) package.
//
// ## Everything here is async
//
// `cryptography_plus`'s entire API surface is `Future`-based (there is a
// `toSync()` escape hatch on some algorithms, but it is not exposed
// uniformly across Ed25519/X25519/AEAD/HKDF, and mixing sync/async call
// styles inside one SDK is its own footgun). Every method in this class is
// therefore `Future`-returning, including `fingerprint` -- the one operation
// every other reference SDK exposes synchronously. This is a deliberate,
// Dart-idiomatic tradeoff: pure non-crypto logic elsewhere in this SDK (CBOR
// codec, timestamp/expiry checks, DNS TXT parsing) stays synchronous, but
// nothing that touches `cryptography_plus` can be.
//
// ## The raw-key-import/export footgun (SimpleKeyPairData / SimplePublicKey)
//
// `cryptography_plus`'s `Ed25519`/`X25519` algorithms both take a raw 32-byte
// seed directly via `newKeyPairFromSeed(seed)` -- no DER/JWK wrapping dance
// needed, unlike Java's JCA or Node's `crypto` module. The resulting
// `SimpleKeyPairData.publicKey` is a `SimplePublicKey` whose `.bytes` are
// already the raw 32-byte wire format this protocol uses. This is the
// friendliest raw-key story of any language in the matrix -- BUT there is
// still a real footgun:
//
// `DartX25519.newKeyPairFromSeed` immediately clamps the seed per RFC 7748
// (clears the low 3 bits, sets bit 254, clears bit 255) and stores the
// CLAMPED bytes as the key pair's private-key bytes -- so
// `extractPrivateKeyBytes()` on an X25519 key pair does NOT return the
// original seed you passed in unless it happened to already be clamped. This
// is harmless for every operation this SDK performs (public-key derivation
// and Diffie-Hellman both re-clamp internally and clamping is idempotent, so
// a fixed conformance-vector private key still derives the correct public
// key and shared secret), but it means an app must never assume
// `localRpIdentityFromBytes(localRpIdentityToBytes(identity))` round-trips
// the exact original X25519 private-key bytes bit-for-bit if it obtained
// them from `extractPrivateKeyBytes()` after key generation elsewhere; this
// SDK always stores exactly the bytes `generateX25519KeyPair` returned, so
// the round trip IS stable for identities this SDK itself generated.
// Verified against `keys.json`'s fixed seeds before building the rest of the
// SDK on this package.
//
// ## Low-order X25519 rejection
//
// Unlike Java's JDK XDH `KeyAgreement` (which throws for several known
// low-order inputs during `doPhase`), `cryptography_plus`'s pure-Dart
// `DartX25519.sharedSecretSync` performs the RFC 7748 scalar multiplication
// unconditionally and returns whatever it computes -- including an all-zero
// shared secret for a malicious all-zero (or other low-order) ephemeral
// public key. The package does NOT reject this itself. This SDK therefore
// adds an explicit all-zero check after every X25519 Diffie-Hellman
// (`x25519DiffieHellman` below), exactly matching
// `callback_box.json`'s `low_order_ephemeral_key_rejected` case.
library;

import 'dart:math';
import 'dart:typed_data';

import 'package:cryptography_plus/cryptography_plus.dart' as cg;

import '../errors.dart';
import 'aead_suite.dart';
import 'hex.dart';

class Ed25519KeyPair {
  final Uint8List publicKey;
  final Uint8List privateKeySeed;
  const Ed25519KeyPair(this.publicKey, this.privateKeySeed);
}

class X25519KeyPair {
  final Uint8List publicKey;
  final Uint8List privateKey;
  const X25519KeyPair(this.publicKey, this.privateKey);
}

class Crypto {
  Crypto._();

  static final cg.Ed25519 _ed25519 = cg.Ed25519();
  static final cg.X25519 _x25519 = cg.X25519();
  static final cg.Sha256 _sha256 = cg.Sha256();
  static final Random _random = Random.secure();

  // -----------------------------------------------------------------
  // Fingerprint / randomness
  // -----------------------------------------------------------------

  /// `sha256(public_key_bytes)`, lowercase hex -- the canonical LinkKeys
  /// fingerprint format, everywhere.
  static Future<String> fingerprint(List<int> publicKeyBytes) async {
    final hash = await _sha256.hash(publicKeyBytes);
    return Hex.encode(hash.bytes);
  }

  /// CSPRNG bytes (`dart:math` `Random.secure()`, the Dart-stdlib equivalent
  /// of `OsRng`/`rand::random`). Synchronous: `Random.secure()` is a plain
  /// stdlib facility, not part of `cryptography_plus`.
  static Uint8List randomBytes(int length) {
    final out = Uint8List(length);
    for (var i = 0; i < length; i++) {
      out[i] = _random.nextInt(256);
    }
    return out;
  }

  static void _requireLen(List<int> b, int len, String what) {
    if (b.length != len) {
      throw CryptoException('$what must be $len bytes, got ${b.length}');
    }
  }

  // -----------------------------------------------------------------
  // Ed25519
  // -----------------------------------------------------------------

  /// Generate a fresh Ed25519 keypair, returning raw 32-byte public key and
  /// 32-byte private seed.
  static Future<Ed25519KeyPair> generateEd25519KeyPair() async {
    final seed = randomBytes(32);
    return ed25519KeyPairFromSeed(seed);
  }

  /// Derive the full keypair (including the public key) for a raw 32-byte
  /// Ed25519 private key seed.
  static Future<Ed25519KeyPair> ed25519KeyPairFromSeed(List<int> seed) async {
    _requireLen(seed, 32, 'Ed25519 private key seed');
    final kp = await _ed25519.newKeyPairFromSeed(seed);
    final data = await kp.extract();
    return Ed25519KeyPair(
      Uint8List.fromList(data.publicKey.bytes),
      Uint8List.fromList(seed),
    );
  }

  /// Sign `message` with an Ed25519 seed (raw 32-byte private key). Returns
  /// a 64-byte signature.
  static Future<Uint8List> signEd25519(
      List<int> message, List<int> privateKeySeed) async {
    _requireLen(privateKeySeed, 32, 'Ed25519 private key seed');
    final kp = await _ed25519.newKeyPairFromSeed(privateKeySeed);
    final sig = await _ed25519.sign(message, keyPair: kp);
    return Uint8List.fromList(sig.bytes);
  }

  /// Verify an Ed25519 signature. Never throws for a malformed
  /// key/signature -- returns `false` uniformly, so callers can treat every
  /// failure mode alike.
  static Future<bool> verifyEd25519(
      List<int> message, List<int> signature, List<int> publicKey) async {
    if (publicKey.length != 32 || signature.length != 64) {
      return false;
    }
    try {
      final sig = cg.Signature(
        signature,
        publicKey: cg.SimplePublicKey(publicKey, type: cg.KeyPairType.ed25519),
      );
      return await _ed25519.verify(message, signature: sig);
    } catch (_) {
      return false;
    }
  }

  // -----------------------------------------------------------------
  // X25519 (ECDH)
  // -----------------------------------------------------------------

  /// Generate a fresh X25519 encryption keypair -- a *separate* key from any
  /// signing key.
  static Future<X25519KeyPair> generateX25519KeyPair() async {
    final kp = await _x25519.newKeyPair();
    final data = await kp.extract();
    final privBytes = await kp.extractPrivateKeyBytes();
    return X25519KeyPair(
      Uint8List.fromList(data.publicKey.bytes),
      Uint8List.fromList(privBytes),
    );
  }

  /// Derive the raw 32-byte X25519 public key for a raw 32-byte private key
  /// (scalar). Needed on the decrypting side of the callback sealed box,
  /// which must feed its OWN public key (not the ephemeral sender's) into
  /// the KDF/AAD construction.
  static Future<Uint8List> derivePublicFromX25519Private(
      List<int> privateKey) async {
    _requireLen(privateKey, 32, 'X25519 private key');
    final kp = await _x25519.newKeyPairFromSeed(privateKey);
    final pub = await kp.extractPublicKey();
    return Uint8List.fromList(pub.bytes);
  }

  /// X25519 Diffie-Hellman, rejecting a non-contributory (low-order) result
  /// (Wire Precision: "reject an all-zero shared secret"). See this
  /// library's docs for why the explicit all-zero check is necessary with
  /// `cryptography_plus` specifically.
  static Future<Uint8List> x25519DiffieHellman(
      List<int> privateKey, List<int> publicKey) async {
    _requireLen(privateKey, 32, 'X25519 private key');
    _requireLen(publicKey, 32, 'X25519 public key');
    try {
      final kp = await _x25519.newKeyPairFromSeed(privateKey);
      final remote = cg.SimplePublicKey(publicKey, type: cg.KeyPairType.x25519);
      final shared =
          await _x25519.sharedSecretKey(keyPair: kp, remotePublicKey: remote);
      final bytes = await shared.extractBytes();
      _rejectLowOrder(bytes);
      return Uint8List.fromList(bytes);
    } on CryptoException {
      rethrow;
    } catch (e) {
      throw CryptoException('X25519 key agreement failed', cause: e);
    }
  }

  static void _rejectLowOrder(List<int> sharedSecret) {
    var allZero = true;
    for (final b in sharedSecret) {
      if (b != 0) {
        allZero = false;
        break;
      }
    }
    if (allZero) {
      throw CryptoException('non-contributory (low-order) X25519 key rejected');
    }
  }

  // -----------------------------------------------------------------
  // HKDF-SHA256
  // -----------------------------------------------------------------

  /// `HKDF-SHA256(salt=none, ikm).expand(info, length)`. `cryptography_plus`'s
  /// `Hkdf.deriveKey` calls its `nonce` parameter the salt.
  ///
  /// Empirically verified (not merely inferred from the RFC): the package's
  /// `DartHmac` throws `ArgumentError` for an EMPTY secret key ("Secret key
  /// must be non-empty"), so the default `nonce: const []` is NOT usable as
  /// a stand-in for "no salt" the way it is in some other HMAC
  /// implementations. This method therefore passes an explicit
  /// hash-length (32-byte) all-zero salt, exactly matching RFC 5869's "if
  /// not provided, salt is set to a string of HashLen zero bytes" -- the
  /// same convention `hkdf::Hkdf::<Sha256>::new(None, ikm)` uses on the Rust
  /// side. Verified against `callback_box.json`'s derived AEAD keys.
  static Future<Uint8List> hkdfSha256(
      List<int> ikm, List<int> info, int length) async {
    final hkdf = cg.Hkdf(hmac: cg.Hmac.sha256(), outputLength: length);
    final zeroSalt = Uint8List(32);
    final out = await hkdf.deriveKey(
        secretKey: cg.SecretKey(ikm), nonce: zeroSalt, info: info);
    return Uint8List.fromList(out.bytes);
  }

  // -----------------------------------------------------------------
  // AEAD dispatch (AES-256-GCM baseline, ChaCha20-Poly1305 optional)
  // -----------------------------------------------------------------

  static const int _aeadTagLength = 16;

  static cg.Cipher _aeadCipher(AeadSuite suite) {
    return switch (suite) {
      AeadSuite.aes256Gcm => cg.AesGcm.with256bits(),
      AeadSuite.chacha20Poly1305 => cg.Chacha20.poly1305Aead(),
    };
  }

  /// Encrypt under `suite`. Output is `ciphertext || 16-byte tag`
  /// (RustCrypto/JCA/Node convention -- this SDK concatenates them itself,
  /// since `cryptography_plus` returns them as separate `SecretBox` fields).
  static Future<Uint8List> aeadEncrypt(AeadSuite suite, List<int> key,
      List<int> nonce, List<int> aad, List<int> plaintext) async {
    _requireLen(key, 32, 'AEAD key');
    _requireLen(nonce, 12, 'AEAD nonce');
    final cipher = _aeadCipher(suite);
    final box = await cipher.encrypt(
      plaintext,
      secretKey: cg.SecretKey(key),
      nonce: nonce,
      aad: aad,
    );
    final out = Uint8List(box.cipherText.length + box.mac.bytes.length);
    out.setAll(0, box.cipherText);
    out.setAll(box.cipherText.length, box.mac.bytes);
    return out;
  }

  /// Decrypt `ciphertext || tag` under `suite`. Throws [CryptoException] on
  /// any authentication failure (tampering, wrong key/nonce/AAD, truncation)
  /// -- never returns unauthenticated plaintext.
  static Future<Uint8List> aeadDecrypt(AeadSuite suite, List<int> key,
      List<int> nonce, List<int> aad, List<int> ciphertextWithTag) async {
    _requireLen(key, 32, 'AEAD key');
    _requireLen(nonce, 12, 'AEAD nonce');
    if (ciphertextWithTag.length < _aeadTagLength) {
      throw CryptoException('ciphertext shorter than the AEAD tag');
    }
    final tagStart = ciphertextWithTag.length - _aeadTagLength;
    final cipherText = ciphertextWithTag.sublist(0, tagStart);
    final tag = ciphertextWithTag.sublist(tagStart);
    try {
      final cipher = _aeadCipher(suite);
      final box = cg.SecretBox(cipherText, nonce: nonce, mac: cg.Mac(tag));
      final plaintext =
          await cipher.decrypt(box, secretKey: cg.SecretKey(key), aad: aad);
      return Uint8List.fromList(plaintext);
    } catch (e) {
      throw CryptoException(
          'AEAD decryption failed (tampering, wrong key, or wrong AAD)',
          cause: e);
    }
  }

  /// Constant-time byte-array equality: fixed-length XOR-accumulate, no
  /// early return on a byte mismatch (only on a length mismatch, which is
  /// not secret data worth hiding). Dart's standard library has no
  /// constant-time compare, unlike e.g. Node's `crypto.timingSafeEqual` or
  /// Go's `crypto/subtle.ConstantTimeCompare` -- `LocalRp.verifyNonceState`
  /// and `LocalRp.checkCallbackHeaderMatchesPayload` use this rather than a
  /// short-circuiting loop for the nonce/state comparisons that gate
  /// CSRF/replay protection, so how many leading bytes matched can't leak
  /// via timing. Signature/MAC verification already goes through
  /// `cryptography_plus`'s own constant-time primitives and doesn't need
  /// this helper.
  static bool constantTimeEquals(List<int> a, List<int> b) {
    if (a.length != b.length) return false;
    var diff = 0;
    for (var i = 0; i < a.length; i++) {
      diff |= a[i] ^ b[i];
    }
    return diff == 0;
  }
}

File: sdks/local-rp/csharp/src/LinkKeys.LocalRp/Crypto/Crypto.cs
using System.Security.Cryptography;
using NSec.Cryptography;

namespace LinkKeys.LocalRp.Crypto;

/// <summary>
/// Crypto primitives this SDK needs (design doc, Language Crypto Matrix,
/// "C#/.NET" row): <a href="https://nsec.rocks/">NSec.Cryptography</a> for
/// Ed25519 sign/verify and X25519 key agreement (the BCL has no Ed25519/X25519
/// support at all — <see cref="ECDiffieHellman"/> is NIST-curves-only), and the
/// BCL for everything else: <see cref="AesGcm"/> and
/// <see cref="System.Security.Cryptography.ChaCha20Poly1305"/> for AEAD,
/// <see cref="HKDF"/> for key derivation, <see cref="SHA256"/> for
/// fingerprinting, <see cref="RandomNumberGenerator"/> for CSPRNG bytes.
///
/// <para><b>Why not NSec's AEAD?</b> NSec's own AES-256-GCM binding is gated on
/// hardware AES-NI support (it throws/is unavailable when the CPU lacks the
/// instruction, per the design doc's language matrix note); the BCL's
/// <see cref="AesGcm"/> goes through OpenSSL on Linux (via
/// <c>System.Security.Cryptography.Native.OpenSsl</c>) and has no such gate.
/// <see cref="System.Security.Cryptography.ChaCha20Poly1305.IsSupported"/> is
/// checked at runtime before use (see <see cref="AeadEncrypt"/>/
/// <see cref="AeadDecrypt"/>); on this box (OpenSSL-backed Linux) it reports
/// supported, so the optional suite is fully exercised rather than skipped.</para>
///
/// <para><b>NSec's low-order X25519 behavior</b>: <see cref="X25519DiffieHellman"/>
/// treats <see cref="KeyAgreementAlgorithm.Agree"/> returning <c>null</c> the
/// same as an explicit all-zero-output check — NSec's libsodium-backed X25519
/// already rejects the all-zero/low-order case at the C layer by returning a
/// null <see cref="SharedSecret"/>, so this method never even sees an all-zero
/// byte array to check; either way, rejection is uniform and verified against
/// <c>callback_box.json</c>'s <c>low_order_ephemeral_key_rejected</c> vector.</para>
/// </summary>
public static class Crypto
{
    private static readonly SignatureAlgorithm Ed25519Algo = SignatureAlgorithm.Ed25519;
    private static readonly KeyAgreementAlgorithm X25519Algo = KeyAgreementAlgorithm.X25519;

    // KeyCreationParameters/SharedSecretCreationParameters are ref structs (NSec), so
    // they cannot be cached in a static field; build a fresh one per call instead.
    private static KeyCreationParameters ExportableKeyParams() =>
        new() { ExportPolicy = KeyExportPolicies.AllowPlaintextExport };

    private static SharedSecretCreationParameters ExportableSecretParams() =>
        new() { ExportPolicy = KeyExportPolicies.AllowPlaintextExport };

    // -----------------------------------------------------------------
    // Fingerprint / randomness
    // -----------------------------------------------------------------

    /// <summary><c>sha256(public_key_bytes)</c>, lowercase hex — the canonical LinkKeys fingerprint format, everywhere.</summary>
    public static string Fingerprint(byte[] publicKeyBytes) => Hex.Encode(SHA256.HashData(publicKeyBytes));

    /// <summary>CSPRNG bytes (<see cref="RandomNumberGenerator"/>).</summary>
    public static byte[] RandomBytes(int length) => RandomNumberGenerator.GetBytes(length);

    private static void RequireLen(byte[] b, int len, string what)
    {
        if (b.Length != len)
        {
            throw new CryptoException($"{what} must be {len} bytes, got {b.Length}");
        }
    }

    // -----------------------------------------------------------------
    // Ed25519
    // -----------------------------------------------------------------

    public sealed record Ed25519KeyPair(byte[] PublicKey, byte[] PrivateKeySeed);

    /// <summary>Generate a fresh Ed25519 keypair, returning raw 32-byte public key and 32-byte private seed.</summary>
    public static Ed25519KeyPair GenerateEd25519KeyPair()
    {
        using var key = Key.Create(Ed25519Algo, ExportableKeyParams());
        byte[] pub = key.PublicKey.Export(KeyBlobFormat.RawPublicKey);
        byte[] seed = key.Export(KeyBlobFormat.RawPrivateKey);
        return new Ed25519KeyPair(pub, seed);
    }

    /// <summary>Raw 32-byte Ed25519 public key -&gt; an NSec <see cref="PublicKey"/> for verification.</summary>
    public static PublicKey ImportEd25519PublicKey(byte[] raw)
    {
        RequireLen(raw, 32, "Ed25519 public key");
        if (!PublicKey.TryImport(Ed25519Algo, raw, KeyBlobFormat.RawPublicKey, out var pub) || pub is null)
        {
            throw new CryptoException("invalid Ed25519 public key bytes");
        }

        return pub;
    }

    /// <summary>Raw 32-byte Ed25519 private key seed -&gt; an NSec <see cref="Key"/> for signing.</summary>
    public static Key ImportEd25519PrivateKey(byte[] seed)
    {
        RequireLen(seed, 32, "Ed25519 private key seed");
        try
        {
            return Key.Import(Ed25519Algo, seed, KeyBlobFormat.RawPrivateKey, ExportableKeyParams());
        }
        catch (Exception e) when (e is FormatException or ArgumentException)
        {
            throw new CryptoException("invalid Ed25519 private key seed", e);
        }
    }

    /// <summary>Sign <paramref name="message"/> with an Ed25519 seed (raw 32-byte private key). Returns a 64-byte signature.</summary>
    public static byte[] SignEd25519(byte[] message, byte[] privateKeySeed)
    {
        using var key = ImportEd25519PrivateKey(privateKeySeed);
        return Ed25519Algo.Sign(key, message);
    }

    /// <summary>
    /// Verify an Ed25519 signature. Never throws for a malformed key/signature — returns
    /// <c>false</c> uniformly, so callers can treat every failure mode alike (design doc:
    /// signature verification is a closed, uniform pass/fail boundary).
    /// </summary>
    public static bool VerifyEd25519(byte[] message, byte[] signature, byte[] publicKey)
    {
        if (publicKey.Length != 32 || signature.Length != 64)
        {
            return false;
        }

        if (!PublicKey.TryImport(Ed25519Algo, publicKey, KeyBlobFormat.RawPublicKey, out var pub) || pub is null)
        {
            return false;
        }

        return Ed25519Algo.Verify(pub, message, signature);
    }

    // -----------------------------------------------------------------
    // X25519 (key agreement)
    // -----------------------------------------------------------------

    public sealed record X25519KeyPair(byte[] PublicKey, byte[] PrivateKey);

    /// <summary>Generate a fresh X25519 encryption keypair — a *separate* key from any signing key.</summary>
    public static X25519KeyPair GenerateX25519KeyPair()
    {
        using var key = Key.Create(X25519Algo, ExportableKeyParams());
        byte[] pub = key.PublicKey.Export(KeyBlobFormat.RawPublicKey);
        byte[] priv = key.Export(KeyBlobFormat.RawPrivateKey);
        return new X25519KeyPair(pub, priv);
    }

    /// <summary>Raw 32-byte X25519 public key (u-coordinate, RFC 7748) -&gt; an NSec <see cref="PublicKey"/>.</summary>
    public static PublicKey ImportX25519PublicKey(byte[] raw)
    {
        RequireLen(raw, 32, "X25519 public key");
        if (!PublicKey.TryImport(X25519Algo, raw, KeyBlobFormat.RawPublicKey, out var pub) || pub is null)
        {
            throw new CryptoException("invalid X25519 public key bytes");
        }

        return pub;
    }

    /// <summary>Raw 32-byte X25519 private scalar -&gt; an NSec <see cref="Key"/>. Clamping happens internally at agreement time.</summary>
    public static Key ImportX25519PrivateKey(byte[] raw)
    {
        RequireLen(raw, 32, "X25519 private key");
        try
        {
            return Key.Import(X25519Algo, raw, KeyBlobFormat.RawPrivateKey, ExportableKeyParams());
        }
        catch (Exception e) when (e is FormatException or ArgumentException)
        {
            throw new CryptoException("invalid X25519 private key bytes", e);
        }
    }

    /// <summary>RFC 7748 base point (<c>u = 9</c>), little-endian 32-byte encoding. Used by <see cref="DerivePublicFromX25519Private"/>.</summary>
    private static readonly byte[] X25519BasePointRaw = BuildBasePoint();

    private static byte[] BuildBasePoint()
    {
        var b = new byte[32];
        b[0] = 9;
        return b;
    }

    /// <summary>
    /// Derive the raw 32-byte X25519 public key for a raw 32-byte private scalar — needed
    /// on the decrypting side of the callback sealed box, which must feed its OWN public
    /// key (not the ephemeral sender's) into the KDF/AAD construction from only its private
    /// key. NSec (like JCA) has no direct "scalar -&gt; point" primitive independent of a key
    /// agreement, but scalar multiplication by the RFC 7748 base point <em>is</em> exactly
    /// what a key agreement against the base point "public key" computes — the same
    /// well-known trick the Java SDK uses, needing zero hand-rolled curve arithmetic.
    /// </summary>
    public static byte[] DerivePublicFromX25519Private(byte[] privateKey) =>
        X25519DiffieHellman(privateKey, X25519BasePointRaw);

    /// <summary>
    /// X25519 Diffie-Hellman, rejecting a non-contributory (low-order) result (Wire
    /// Precision: "reject an all-zero shared secret"). NSec's libsodium-backed
    /// <see cref="KeyAgreementAlgorithm.Agree"/> already rejects known low-order inputs by
    /// returning <c>null</c> rather than an all-zero secret; this method treats a null result
    /// identically to an explicit all-zero check, so rejection is uniform regardless of which
    /// layer catches it.
    /// </summary>
    public static byte[] X25519DiffieHellman(byte[] privateKey, byte[] publicKey)
    {
        using var priv = ImportX25519PrivateKey(privateKey);
        var pub = ImportX25519PublicKey(publicKey);
        using var shared = X25519Algo.Agree(priv, pub, ExportableSecretParams());
        if (shared is null)
        {
            throw new CryptoException("non-contributory (low-order) X25519 key rejected");
        }

        byte[] sharedBytes = shared.Export(SharedSecretBlobFormat.RawSharedSecret);
        if (IsAllZero(sharedBytes))
        {
            throw new CryptoException("non-contributory (low-order) X25519 key rejected");
        }

        return sharedBytes;
    }

    private static bool IsAllZero(byte[] b)
    {
        foreach (var x in b)
        {
            if (x != 0)
            {
                return false;
            }
        }

        return true;
    }

    // -----------------------------------------------------------------
    // HKDF-SHA256 (BCL: System.Security.Cryptography.HKDF)
    // -----------------------------------------------------------------

    /// <summary>
    /// HKDF-SHA256(salt=none, ikm).expand(info, length) — exactly
    /// <c>hkdf::Hkdf::&lt;Sha256&gt;::new(None, ikm).expand(info, &amp;mut out)</c> on the Rust
    /// side. A <c>null</c>/empty salt is treated by <see cref="HKDF.DeriveKey"/> as a
    /// zero-filled salt of hash length per RFC 5869, matching every other SDK's HKDF call.
    /// </summary>
    public static byte[] HkdfSha256(byte[] ikm, byte[] info, int length) =>
        HKDF.DeriveKey(HashAlgorithmName.SHA256, ikm, length, salt: null, info: info);

    // -----------------------------------------------------------------
    // AEAD dispatch (AES-256-GCM baseline, ChaCha20-Poly1305 optional)
    // -----------------------------------------------------------------

    private const int AeadTagLen = 16;
    private const int AeadNonceLen = 12;

    /// <summary>Whether <paramref name="suite"/> is usable on this runtime. Always true for AES-256-GCM.</summary>
    public static bool IsSuiteSupported(AeadSuite suite) => suite switch
    {
        AeadSuite.Aes256Gcm => AesGcm.IsSupported,
        AeadSuite.ChaCha20Poly1305 => System.Security.Cryptography.ChaCha20Poly1305.IsSupported,
        _ => false,
    };

    /// <summary>Encrypt under <paramref name="suite"/>. Output is <c>ciphertext || 16-byte tag</c> (the wire convention every SDK uses).</summary>
    public static byte[] AeadEncrypt(AeadSuite suite, byte[] key, byte[] nonce, byte[] aad, byte[] plaintext)
    {
        RequireLen(key, 32, "AEAD key");
        RequireLen(nonce, AeadNonceLen, "AEAD nonce");
        RequireSuiteSupported(suite);

        var ciphertext = new byte[plaintext.Length];
        var tag = new byte[AeadTagLen];
        try
        {
            switch (suite)
            {
                case AeadSuite.Aes256Gcm:
                    using (var aes = new AesGcm(key, AeadTagLen))
                    {
                        aes.Encrypt(nonce, plaintext, ciphertext, tag, aad);
                    }

                    break;
                case AeadSuite.ChaCha20Poly1305:
                    using (var cc = new System.Security.Cryptography.ChaCha20Poly1305(key))
                    {
                        cc.Encrypt(nonce, plaintext, ciphertext, tag, aad);
                    }

                    break;
                default:
                    throw new ArgumentOutOfRangeException(nameof(suite));
            }
        }
        catch (CryptographicException e)
        {
            throw new CryptoException("AEAD encryption failed", e);
        }

        var combined = new byte[ciphertext.Length + tag.Length];
        ciphertext.CopyTo(combined, 0);
        tag.CopyTo(combined, ciphertext.Length);
        return combined;
    }

    /// <summary>
    /// Decrypt <c>ciphertext || tag</c> under <paramref name="suite"/>. Throws
    /// <see cref="CryptoException"/> on any authentication failure (tampering, wrong
    /// key/nonce/AAD, truncation) — never returns unauthenticated plaintext.
    /// </summary>
    public static byte[] AeadDecrypt(AeadSuite suite, byte[] key, byte[] nonce, byte[] aad, byte[] ciphertextAndTag)
    {
        RequireLen(key, 32, "AEAD key");
        RequireLen(nonce, AeadNonceLen, "AEAD nonce");
        RequireSuiteSupported(suite);

        if (ciphertextAndTag.Length < AeadTagLen)
        {
            throw new CryptoException("AEAD decryption failed (ciphertext shorter than the authentication tag)");
        }

        int ctLen = ciphertextAndTag.Length - AeadTagLen;
        var ciphertext = ciphertextAndTag.AsSpan(0, ctLen);
        var tag = ciphertextAndTag.AsSpan(ctLen, AeadTagLen);
        var plaintext = new byte[ctLen];

        try
        {
            switch (suite)
            {
                case AeadSuite.Aes256Gcm:
                    using (var aes = new AesGcm(key, AeadTagLen))
                    {
                        aes.Decrypt(nonce, ciphertext, tag, plaintext, aad);
                    }

                    break;
                case AeadSuite.ChaCha20Poly1305:
                    using (var cc = new System.Security.Cryptography.ChaCha20Poly1305(key))
                    {
                        cc.Decrypt(nonce, ciphertext, tag, plaintext, aad);
                    }

                    break;
                default:
                    throw new ArgumentOutOfRangeException(nameof(suite));
            }
        }
        catch (CryptographicException e)
        {
            throw new CryptoException("AEAD decryption failed (tampering, wrong key, or wrong AAD)", e);
        }

        return plaintext;
    }

    private static void RequireSuiteSupported(AeadSuite suite)
    {
        if (!IsSuiteSupported(suite))
        {
            throw new CryptoException($"AEAD suite {suite.WireId()} is not supported on this runtime");
        }
    }

    /// <summary>Constant-time byte-array equality (defense in depth; NSec/BCL's own MAC/signature checks are already constant-time).</summary>
    public static bool ConstantTimeEquals(byte[] a, byte[] b) => CryptographicOperations.FixedTimeEquals(a, b);
}
